File: Sources/Scheduler.swift
import Foundation

/// Core scheduling algorithm for Task Tracker.
/// Works backwards from each task's deadline to determine if it's achievable,
/// then builds a day-by-day schedule across all tasks.
struct Scheduler {

    // MARK: - Public API

    /// Generates a schedule for all given tasks.
    /// - Parameters:
    ///   - tasks: List of tasks to schedule
    ///   - settings: Daily work hour settings
    /// - Returns: ScheduleResult with day-by-day blocks and any impossible tasks
    static func generateSchedule(tasks: [Task], settings: DailySettings) -> ScheduleResult {
        guard !tasks.isEmpty else { return .empty }

        // Sort tasks by deadline (earliest first)
        let sortedTasks = tasks.sorted { $0.deadline < $1.deadline }

        var impossibleTasks: [ImpossibleTask] = []
        // Maps (taskId, date) -> hours allocated
        var allocation: [String: Double] = [:]

        for task in sortedTasks {
            let result = scheduleTask(task, settings: settings, existingAllocations: allocation)
            if result.isPossible {
                for (key, hours) in result.allocations {
                    allocation[key] = hours
                }
            } else {
                impossibleTasks.append(ImpossibleTask(
                    id: task.id,
                    taskTitle: task.title,
                    hoursShort: result.hoursShort
                ))
            }
        }

        // Build day schedules from allocations
        let daySchedules = buildDaySchedules(from: allocation, tasks: sortedTasks, settings: settings)

        return ScheduleResult(daySchedules: daySchedules, impossibleTasks: impossibleTasks)
    }

    // MARK: - Private

    private struct AllocationResult {
        let isPossible: Bool
        let hoursShort: Double
        let allocations: [String: Double] // "\(taskId)_\(dateStr)" -> hours
    }

    private static func scheduleTask(_ task: Task, settings: DailySettings, existingAllocations: [String: Double]) -> AllocationResult {
        let calendar = Calendar.current
        let today = calendar.startOfDay(for: Date())
        let deadline = calendar.startOfDay(for: task.deadline)

        // Collect days from today to deadline (inclusive)
        var days: [Date] = []
        var cursor = deadline
        while cursor >= today {
            days.append(cursor)
            cursor = calendar.date(byAdding: .day, value: -1, to: cursor)!
        }
        if !days.contains(today) {
            days.append(today)
            days.sort()
        }

        let normalPerDay = settings.normalWorkHoursPerDay
        let overtimePerDay = settings.overtimeMaxHoursPerDay
        var remainingHours = task.estimatedHours
        var newAllocations: [String: Double] = [:]
        var usedOvertime = false

        for day in days {
            // Skip days in the past relative to today
            if day < today { continue }

            let dayKey = dayKey(day)
            let existingOnDay = existingAllocationsFor(day: day, taskId: task.id, existing: existingAllocations)

            // Normal hours available (reduced by what's already allocated to this task)
            let normalAvailable = max(0, normalPerDay - existingOnDay)
            var toAllocate = min(normalAvailable, remainingHours)

            if toAllocate > 0 {
                let key = allocationKey(task.id, day)
                newAllocations[key] = toAllocate
                remainingHours -= toAllocate
                if remainingHours <= 0 { break }
            }

            // Overtime hours
            if remainingHours > 0 && settings.overtimeMaxHoursPerDay > 0 {
                let overtimeKey = allocationKey(task.id, day) + "_ot"
                let existingOT = existingAllocations[overtimeKey] ?? 0
                let otAvailable = max(0, overtimePerDay - existingOT)
                var otAlloc = min(otAvailable, remainingHours)
                if otAlloc > 0 {
                    newAllocations[overtimeKey] = otAlloc
                    remainingHours -= otAlloc
                    usedOvertime = true
                    if remainingHours <= 0 { break }
                }
            }
        }

        if remainingHours > 0 {
            return AllocationResult(isPossible: false, hoursShort: remainingHours, allocations: [:])
        }

        return AllocationResult(isPossible: true, hoursShort: 0, allocations: newAllocations)
    }

    private static func buildDaySchedules(from allocation: [String: Double], tasks: [Task], settings: DailySettings) -> [DaySchedule] {
        let calendar = Calendar.current
        let today = calendar.startOfDay(for: Date())

        // Find all unique dates in allocation
        var allDates: Set<String> = []
        for key in allocation.keys {
            let components = key.split(separator: "_")
            if components.count >= 2 {
                let dateStr = components[components.count - 1]
                if !dateStr.hasSuffix("_ot") {
                    allDates.insert(String(dateStr))
                }
            }
        }

        var schedules: [DaySchedule] = []

        for dateStr in allDates.sorted() {
            guard let date = parseDay(dateStr), date >= today else { continue }

            var blocks: [ScheduledBlock] = []
            for task in tasks {
                // Normal block
                let normalKey = allocationKey(task.id, date)
                if let hours = allocation[normalKey], hours > 0 {
                    blocks.append(ScheduledBlock(taskTitle: task.title, taskId: task.id, hours: hours, isOvertime: false))
                }
                // Overtime block
                let otKey = normalKey + "_ot"
                if let hours = allocation[otKey], hours > 0 {
                    blocks.append(ScheduledBlock(taskTitle: task.title, taskId: task.id, hours: hours, isOvertime: true))
                }
            }

            if !blocks.isEmpty {
                schedules.append(DaySchedule(date: date, blocks: blocks))
            }
        }

        return schedules.sorted { $0.date < $1.date }
    }

    private static func existingAllocationsFor(day: Date, taskId: UUID, existing: [String: Double]) -> Double {
        let dayKey = dayKey(day)
        let normalKey = allocationKey(taskId, day)
        let otKey = normalKey + "_ot"
        return (existing[normalKey] ?? 0) + (existing[otKey] ?? 0)
    }

    private static func dayKey(_ date: Date) -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter.string(from: date)
    }

    private static func parseDay(_ str: String) -> Date? {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter.date(from: str)
    }

    private static func allocationKey(_ taskId: UUID, _ date: Date) -> String {
        "\(taskId.uuidString)_\(dayKey(date))"
    }
}

File: scheduler.py
"""
Task Tracker — Scheduling Algorithm (Python port for testing)
"""
from dataclasses import dataclass, field
from datetime import date, datetime, timedelta
from typing import Optional


@dataclass
class Task:
    id: str
    title: str
    estimated_hours: float
    deadline: date
    normal_hours_per_day: float = 8.0


@dataclass
class DailySettings:
    normal_work_hours_per_day: float = 8.0
    overtime_max_hours_per_day: float = 2.0


@dataclass
class ScheduledBlock:
    task_title: str
    task_id: str
    hours: float
    is_overtime: bool


@dataclass
class DaySchedule:
    date: date
    blocks: list[ScheduledBlock] = field(default_factory=list)

    @property
    def is_overtime(self) -> bool:
        return any(b.is_overtime for b in self.blocks)


@dataclass
class ImpossibleTask:
    task_title: str
    hours_short: float


@dataclass
class ScheduleResult:
    day_schedules: list[DaySchedule]
    impossible_tasks: list[ImpossibleTask]

    @staticmethod
    def empty():
        return ScheduleResult(day_schedules=[], impossible_tasks=[])


def generate_schedule(tasks: list[Task], settings: DailySettings) -> ScheduleResult:
    """
    Core scheduling algorithm.
    Works backwards from each task's deadline to determine achievability,
    then builds a day-by-day schedule across all tasks.
    """
    if not tasks:
        return ScheduleResult.empty()

    today = datetime.now().date()
    sorted_tasks = sorted(tasks, key=lambda t: t.deadline)

    impossible_tasks: list[ImpossibleTask] = []
    # allocation: dict[str, float] — key: "taskId_YYYY-MM-DD[_ot]" → hours
    allocation: dict[str, float] = {}

    for task in sorted_tasks:
        result = _schedule_task(task, today, settings, allocation)
        if result["is_possible"]:
            for key, hours in result["allocations"].items():
                allocation[key] = hours
        else:
            impossible_tasks.append(ImpossibleTask(
                task_title=task.title,
                hours_short=result["hours_short"]
            ))

    day_schedules = _build_day_schedules(allocation, sorted_tasks, today)
    return ScheduleResult(day_schedules=day_schedules, impossible_tasks=impossible_tasks)


def _schedule_task(task: Task, today: date, settings: DailySettings, existing: dict[str, float]) -> dict:
    """Schedule a single task. Returns {is_possible, hours_short, allocations}."""
    deadline = task.deadline

    # Collect days from today to deadline (inclusive), but only future days
    days = []
    cursor = deadline
    while cursor >= today:
        if cursor >= today:
            days.append(cursor)
        cursor -= timedelta(days=1)

    if not days:
        # Deadline is in the past
        return {"is_possible": False, "hours_short": task.estimated_hours, "allocations": {}}

    days = sorted(days)  # chronological
    normal_per_day = settings.normal_work_hours_per_day
    overtime_per_day = settings.overtime_max_hours_per_day
    remaining = task.estimated_hours
    new_allocs: dict[str, float] = {}

    for day in days:
        if day < today:
            continue

        day_str = _day_key(day)
        existing_on_day = _existing_on_day(task.id, day, existing)

        # Normal hours
        normal_available = max(0, normal_per_day - existing_on_day)
        to_alloc = min(normal_available, remaining)
        if to_alloc > 0:
            key = f"{task.id}_{day_str}"
            new_allocs[key] = to_alloc
            remaining -= to_alloc
            if remaining <= 0:
                break

        # Overtime hours
        if remaining > 0 and overtime_per_day > 0:
            ot_key = f"{task.id}_{day_str}_ot"
            existing_ot = existing.get(ot_key, 0)
            ot_available = max(0, overtime_per_day - existing_ot)
            ot_alloc = min(ot_available, remaining)
            if ot_alloc > 0:
                new_allocs[ot_key] = ot_alloc
                remaining -= ot_alloc
                if remaining <= 0:
                    break

    if remaining > 0:
        return {"is_possible": False, "hours_short": remaining, "allocations": {}}

    return {"is_possible": True, "hours_short": 0, "allocations": new_allocs}


def _build_day_schedules(allocation: dict[str, float], tasks: list[Task], today: date) -> list[DaySchedule]:
    """Build day-by-day schedule cards from allocation dict."""
    # Find all unique dates
    date_keys: set[str] = set()
    for key in allocation:
        parts = key.split("_")
        if len(parts) >= 2 and not parts[-1] == "ot":
            date_keys.add(parts[-1])

    schedules: list[DaySchedule] = []
    task_map = {t.id: t for t in tasks}

    for dk in sorted(date_keys):
        d = _parse_day(dk)
        if d is None or d < today:
            continue

        blocks: list[ScheduledBlock] = []
        for task_id, task in task_map.items():
            normal_key = f"{task_id}_{dk}"
            normal_hours = allocation.get(normal_key, 0)
            if normal_hours > 0:
                blocks.append(ScheduledBlock(
                    task_title=task.title,
                    task_id=task_id,
                    hours=normal_hours,
                    is_overtime=False
                ))
            ot_hours = allocation.get(f"{normal_key}_ot", 0)
            if ot_hours > 0:
                blocks.append(ScheduledBlock(
                    task_title=task.title,
                    task_id=task_id,
                    hours=ot_hours,
                    is_overtime=True
                ))

        if blocks:
            schedules.append(DaySchedule(date=d, blocks=blocks))

    return sorted(schedules, key=lambda s: s.date)


def _existing_on_day(task_id: str, day: date, existing: dict[str, float]) -> float:
    day_str = _day_key(day)
    return existing.get(f"{task_id}_{day_str}", 0) + existing.get(f"{task_id}_{day_str}_ot", 0)


def _day_key(d: date) -> str:
    return d.isoformat()


def _parse_day(s: str) -> Optional[date]:
    return date.fromisoformat(s)
